File: src/main/java/io/github/biligoldenwater/midiplayer/utils/particlegui/PixelColor.java
package io.github.biligoldenwater.midiplayer.utils.particlegui;

import org.bukkit.Color;

public class PixelColor {
    private double red;
    private double green;
    private double blue;
    private double alpha;

    public PixelColor() {
        this.red = 0;
        this.green = 0;
        this.blue = 0;
        this.alpha = 1;
    }

    public PixelColor(double red, double green, double blue, double alpha) {
        this.red = red;
        this.green = green;
        this.blue = blue;
        this.alpha = alpha;
    }

    public PixelColor(double red, double green, double blue) {
        this(red, green, blue, 1);
    }

    public PixelColor(Color color) {
        this(color.getRed() / 255.0, color.getGreen() / 255.0, color.getBlue() / 255.0);
    }

    public PixelColor overlap(PixelColor pixelColor) {
        if (pixelColor.alpha == 1) {
            this.red = pixelColor.red;
            this.green = pixelColor.green;
            this.blue = pixelColor.blue;
            return this;
        }
        this.red = this.red * (1 - pixelColor.alpha) + pixelColor.red * pixelColor.alpha;
        this.green = this.green * (1 - pixelColor.alpha) + pixelColor.green * pixelColor.alpha;
        this.blue = this.blue * (1 - pixelColor.alpha) + pixelColor.blue * pixelColor.alpha;
        return this;
    }

    public boolean equals(PixelColor pixelColor) {
        return this.red == pixelColor.red && this.green == pixelColor.green && this.blue == pixelColor.blue && this.alpha == pixelColor.alpha;
    }

    public Color toBukkitColor() {
        return Color.fromRGB((int) (red * 255), (int) (green * 255), (int) (blue * 255));
    }

    public double getRed() {
        return red;
    }

    public void setRed(double red) {
        this.red = red;
    }

    public double getGreen() {
        return green;
    }

    public void setGreen(double green) {
        this.green = green;
    }

    public double getBlue() {
        return blue;
    }

    public void setBlue(double blue) {
        this.blue = blue;
    }

    public double getAlpha() {
        return alpha;
    }

    public void setAlpha(double alpha) {
        this.alpha = alpha;
    }
}

File: src/main/java/io/github/biligoldenwater/midiplayer/utils/particlegui/PixelColor.kt
package io.github.biligoldenwater.midiplayer.utils.particlegui

import org.bukkit.Color

class PixelColor {
    var red = 0.0
    var green = 0.0
    var blue = 0.0
    var alpha = 0.0

    constructor() {
        red = 0.0
        green = 0.0
        blue = 0.0
        alpha = 1.0
    }

    @JvmOverloads
    constructor(red: Double, green: Double, blue: Double, alpha: Double = 1.0) {
        this.red = red
        this.green = green
        this.blue = blue
        this.alpha = alpha
    }

    constructor(color: Color) : this(color.red / 255.0, color.green / 255.0, color.blue / 255.0) {}

    fun overlap(pixelColor: PixelColor): PixelColor {
        if (pixelColor.alpha == 1.0) {
            red = pixelColor.red
            green = pixelColor.green
            blue = pixelColor.blue
            return this
        }
        red = red * (1 - pixelColor.alpha) + pixelColor.red * pixelColor.alpha
        green = green * (1 - pixelColor.alpha) + pixelColor.green * pixelColor.alpha
        blue = blue * (1 - pixelColor.alpha) + pixelColor.blue * pixelColor.alpha
        return this
    }

    fun equals(pixelColor: PixelColor): Boolean {
        return red == pixelColor.red && green == pixelColor.green && blue == pixelColor.blue && alpha == pixelColor.alpha
    }

    fun toBukkitColor(): Color {
        return Color.fromRGB((red * 255).toInt(), (green * 255).toInt(), (blue * 255).toInt())
    }
}